sklearn/decomposition/_fastica.py

```diff
@@ -557,9 +618,33 @@ def g(x, fun_args):
             XT -= X_mean[:, np.newaxis]
 
             # Whitening and preprocessing by PCA
-            u, d, _ = linalg.svd(XT, full_matrices=False, check_finite=False)
+            if self.whiten_solver == "eigh":
+                # Faster when num_samples >> n_features
+                d, u = linalg.eigh(XT.dot(X))
+                sort_indices = np.argsort(d)[::-1]
+                eps = np.finfo(d.dtype).eps
+                degenerate_idx = d < eps
+                if np.any(degenerate_idx):
+                    warnings.warn(
+                        "There are some small singular values, using "
+                        "whiten_solver = 'svd' might lead to more "
+                        "accurate results."
+                    )
+                d[degenerate_idx] = eps  # For numerical issues
+                np.sqrt(d, out=d)
+                d, u = d[sort_indices], u[:, sort_indices]
+            elif self.whiten_solver == "svd":
+                u, d = linalg.svd(XT, full_matrices=False, check_finite=False)[:2]
+            else:
+                raise ValueError(
+                    "`whiten_solver` must be 'eigh' or 'svd' but got"
+                    f" {self.whiten_solver} instead"
+                )
+
+            # Give consistent eigenvectors for both svd solvers
+            if self.sign_flip:
+                u *= np.sign(u[0])
```

Comment:
I'm too late on this one but it this be worth using the "standardized" svd_flip function used in other similar situations ?

https://github.com/scikit-learn/scikit-learn/blob/4b3a69aa2b6d54203795e20a9a68d0b93e4da997/sklearn/utils/extmath.py#L895